src/conduits/RwaInputConduit3.sol

```diff
@@ -0,0 +1,310 @@
+        may[usr] = 1;
+        emit Mate(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` push access from this contract.
+     * @param usr The user address.
+     */
+    function hate(address usr) external auth {
+        may[usr] = 0;
+        emit Hate(usr);
+    }
+
+    /*//////////////////////////////////
+               Administration
+    //////////////////////////////////*/
+
+    /**
+     * @notice Updates a contract parameter.
+     * @param what The changed parameter name. `"to", "quitTo"`
+     * @param data The new value of the parameter.
+     */
+    function file(bytes32 what, address data) external auth {
+        if (what == "quitTo") {
+            quitTo = data;
+        } else if (what == "to") {
+            to = data;
+        } else {
+            revert("RwaInputConduit3/unrecognised-param");
+        }
+
+        emit File(what, data);
+    }
+
+    /*//////////////////////////////////
+               Operations
+    //////////////////////////////////*/
+
+    /**
+     * @notice Swaps the GEM balance of this contract into DAI through the PSM and push it into the `to` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     */
+    function push() external isMate {
+        _doPush(gem.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Swaps the specified amount of GEM into DAI through the PSM and push it into the `to` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     * @param amt Gem amount.
+     */
+    function push(uint256 amt) external isMate {
+        _doPush(amt);
+    }
+
+    /**
+     * @notice Flushes out any GEM balance to `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     */
+    function quit() external isMate {
+        _doQuit(gem.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Flushes out specific amount of GEM balance to `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     * @param amt Gem amount.
+     */
+    function quit(uint256 amt) external isMate {
+        _doQuit(amt);
+    }
+
+    /**
+     * @notice Flushes out all outstanding `token` balance to `usr` address.
+     * @dev Can only be called by the admin
+     * @param token Destination address.
+     * @param usr Destination address.
+     * @param amt Token amount.
+     */
+    function yank(
+        address token,
+        address usr,
+        uint256 amt
+    ) external auth {
+        GemAbstract(token).transfer(usr, amt);
+        emit Yank(token, usr, amt);
+    }
+
+    /**
+     * @notice Calculate required amount of GEM to get `wad` amount of DAI.
+     * @param wad DAI amount.
+     * @return gemAmt Amount of GEM required.
+     */
+    function requiredGemAmt(uint256 wad) external view returns (uint256 gemAmt) {
+        return mul(wad, WAD) / mul(sub(WAD, psm.tin()), to18ConvertionFactor);
```

Comment:
Probably better to do the math in `WAD` units and only do the division of `to18ConvertionFactor` at the end, i.e:
`return (mul(wad, WAD) / sub(WAD, psm.tin())) / to18ConvertionFactor;`

This way if to18ConvertionFactor is relatively large we will lose less percision.

Also, this function seems untested.

Reply:
Ah I see something similar was discussed here - https://github.com/makerdao/mip21-toolkit/pull/21#discussion_r970816929
Will resolve for now and discuss with @gbalabasquer separately.